src/cargo/core/compiler/mod.rs

```diff
@@ -589,15 +591,26 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
     if let CompileKind::Target(target) = unit.kind {
         rustdoc.arg("--target").arg(target.rustc_target());
     }
-
     let doc_dir = cx.files().out_dir(unit);
+    // We need to make sure that if there were any previous docs
+    // already compiled, they were compiled with the same Rustc version that we're currently
+    // using. Otherways we must remove the `doc/` folder and compile again forcing a rebuild.
+    //
+    // This is important because the `.js`/`.html` & `.css` files that are generated by Rustc don't have
+    // any versioning (See https://github.com/rust-lang/cargo/issues/8461).
+    // Therefore, we can end up with weird bugs and behaviours if we mix different
+    // compiler versions of these files.
+    let (fingerprint, doc_dir_removed) = RustDocFingerprint::check_rustdoc_fingerprint(
```

Comment:
I added it here since there's [this comment](https://github.com/rust-lang/cargo/blob/de4230293d01000fe13052e9bef38278d7b268fd/src/cargo/core/compiler/mod.rs#L595) where the doc folder is created. Therefore I assumed this would be the best place to check wether we want to remove it or not etc..

Reply:
Also, it's important to highlight that inside of the `Worker` we only write the fingerprint in case we erased the previous one. But the directory removal/fingerprint reading is done outside. So I don't really know if it is unsafe (obviously, you're the expert here) but just writing the fingerprint in case it was previously erased inside of the `Worker` is unsafe?

Reply:
This is called for every package that is being documented, so we probably don't want it to be called N times.  By lifting it out to an earlier stage, that can help ensure that it only runs once.